cli/internal/run/run.go

```diff
@@ -311,161 +274,84 @@ func (r *run) runOperation(ctx gocontext.Context, g *graph.CompleteGraph, rs *ru
-			}
-			if err != nil {
-				return err
-			}
-			r.base.UI.Output(rendered)
-		} else {
-			if err := displayDryTextRun(r.base.UI, tasksRun, packagesInScope, g.PackageInfos, r.opts.runOpts.singlePackage); err != nil {
-				return err
-			}
-		}
-	} else {
-		packagesInScope := rs.FilteredPkgs.UnsafeListOfStrings()
-		sort.Strings(packagesInScope)
-		if r.opts.runOpts.singlePackage {
-			r.base.UI.Output(fmt.Sprintf("%s %s", ui.Dim("• Running"), ui.Dim(ui.Bold(strings.Join(rs.Targets, ", ")))))
-		} else {
-			r.base.UI.Output(fmt.Sprintf(ui.Dim("• Packages in scope: %v"), strings.Join(packagesInScope, ", ")))
-			r.base.UI.Output(fmt.Sprintf("%s %s %s", ui.Dim("• Running"), ui.Dim(ui.Bold(strings.Join(rs.Targets, ", "))), ui.Dim(fmt.Sprintf("in %v packages", rs.FilteredPkgs.Len()))))
-		}
-		return r.executeTasks(ctx, g, rs, engine, packageManager, tracker, startAt)
+		return GraphRun(ctx, rs, engine, r.base)
 	}
-	return nil
-}
 
-func renderDryRunSinglePackageJSON(tasksRun []hashedTask) (string, error) {
-	singlePackageTasks := make([]hashedSinglePackageTask, len(tasksRun))
-	for i, ht := range tasksRun {
-		singlePackageTasks[i] = ht.toSinglePackageTask()
-	}
-	dryRun := &struct {
-		Tasks []hashedSinglePackageTask `json:"tasks"`
-	}{singlePackageTasks}
-	bytes, err := json.MarshalIndent(dryRun, "", "  ")
-	if err != nil {
-		return "", errors.Wrap(err, "failed to render JSON")
-	}
-	return string(bytes), nil
-}
+	packagesInScope := rs.FilteredPkgs.UnsafeListOfStrings()
+	sort.Strings(packagesInScope)
+	// Initiate analytics and cache
+	analyticsClient := r.initAnalyticsClient(ctx)
+	defer analyticsClient.CloseWithTimeout(50 * time.Millisecond)
+	turboCache, err := r.initCache(ctx, rs, analyticsClient)
 
-func renderDryRunFullJSON(tasksRun []hashedTask, packagesInScope []string) (string, error) {
-	dryRun := &struct {
-		Packages []string     `json:"packages"`
-		Tasks    []hashedTask `json:"tasks"`
-	}{
-		Packages: packagesInScope,
-		Tasks:    tasksRun,
-	}
-	bytes, err := json.MarshalIndent(dryRun, "", "  ")
 	if err != nil {
-		return "", errors.Wrap(err, "failed to render JSON")
-	}
-	return string(bytes), nil
-}
-
-func displayDryTextRun(ui cli.Ui, tasksRun []hashedTask, packagesInScope []string, packageInfos map[interface{}]*fs.PackageJSON, isSinglePackage bool) error {
-	if !isSinglePackage {
-		ui.Output("")
-		ui.Info(util.Sprintf("${CYAN}${BOLD}Packages in Scope${RESET}"))
-		p := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
-		fmt.Fprintln(p, "Name\tPath\t")
-		for _, pkg := range packagesInScope {
-			fmt.Fprintf(p, "%s\t%s\t\n", pkg, packageInfos[pkg].Dir)
-		}
-		if err := p.Flush(); err != nil {
-			return err
+		if errors.Is(err, cache.ErrNoCachesEnabled) {
+			r.base.UI.Warn("No caches are enabled. You can try \"turbo login\", \"turbo link\", or ensuring you are not passing --remote-only to enable caching")
+		} else {
+			return errors.Wrap(err, "failed to set up caching")
```

Comment:
I think the refactor is correct, but if I'm reading the existing code we would call `turboCache.Shutdown()` via defer when this codepath got hit.

Reply:
By "existing code", you mean in this branch or in `main`?

In `main`, `turboCache.Shutdown()` would _not_ happen if we hit this error, it would return out before then.

https://github.com/vercel/turbo/blob/335de9c72ec344e5d262071ab2d9c1e7ce830727/cli/internal/run/run.go#L757-L767

Reply:
Sorry, should've been more clear, I meant stuff in `main` and I was referring to the dry execution codepath (my brain was fried when I started reviewing this 🙃 ):
https://github.com/vercel/turbo/blob/335de9c72ec344e5d262071ab2d9c1e7ce830727/cli/internal/run/run.go#L875-L886